File: core/report.py
"""Результаты диагностики."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone

from core.status import TestStatus, overall_from_tests
from core.stats import looks_like_square_gate
from pad_common import APP_NAME, REPORT_SCHEMA, VERSION


@dataclass
class StickTestResult:
    status: TestStatus = TestStatus.NOT_TESTED
    drift_pct: float = 0.0
    mean_x: float = 0.0
    mean_y: float = 0.0
    max_radius: float = 0.0
    deadzone_pct: float = 0.0
    physical_drift_pct: float = 0.0
    circularity_pct: float = 0.0
    range_ok: bool = True
    issues: list[str] = field(default_factory=list)
    rest_points: list[tuple[float, float]] = field(default_factory=list)


@dataclass
class TriggerTestResult:
    status: TestStatus = TestStatus.NOT_TESTED
    min_value: float = 0.0
    max_value: float = 0.0
    spike_count: int = 0
    returns_to_zero: bool = True
    issues: list[str] = field(default_factory=list)
    timeline: list[tuple[int, float]] = field(default_factory=list)


@dataclass
class ButtonCheck:
    index: int
    name: str
    pressed: bool = False
    held: bool = False
    skipped: bool = False
    sticky: bool = False
    sensitive: bool = True
    tap_count: int = 0
    issues: list[str] = field(default_factory=list)


@dataclass
class ButtonsTestResult:
    status: TestStatus = TestStatus.NOT_TESTED
    buttons: list[ButtonCheck] = field(default_factory=list)
    pressed_count: int = 0
    held_count: int = 0
    issues: list[str] = field(default_factory=list)


@dataclass
class EventRateResult:
    mean_interval_ms: float = 0.0
    median_interval_ms: float = 0.0
    max_interval_ms: float = 0.0
    estimated_hz: float = 0.0
    note: str = ""


def compute_score(report: "DiagnosticReport") -> int:
    score = 10.0
    for stick in (report.left_stick, report.right_stick):
        if stick.status == TestStatus.FAIL:
            score -= 3.0
        elif stick.status == TestStatus.WARN:
            score -= 1.5
    for trig in (report.lt, report.rt):
        if trig.status == TestStatus.FAIL:
            score -= 2.0
        elif trig.status == TestStatus.WARN:
            score -= 0.75
    buttons = report.buttons
    if buttons.status not in (TestStatus.NOT_TESTED, TestStatus.NOT_SUPPORTED):
        missed = sum(1 for item in buttons.buttons if not item.pressed or item.skipped)
        not_held = sum(1 for item in buttons.buttons if item.pressed and not item.held and not item.skipped)
        sticky = sum(1 for item in buttons.buttons if item.sticky)
        dull = sum(1 for item in buttons.buttons if not item.sensitive)
        score -= min(6.0, missed * 1.5)
        score -= min(2.0, not_held * 0.5)
        score -= min(3.0, sticky * 1.0)
        score -= min(3.0, dull * 1.0)
    return max(1, min(10, int(round(score))))


def score_tone(score: int) -> str:
    if score >= 8:
        return "good"
    if score >= 5:
        return "ok"
    return "bad"


@dataclass
class DiagnosticReport:
    schema_version: str = REPORT_SCHEMA
    app_name: str = APP_NAME
    app_version: str = VERSION
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    locale: str = "ru"
    device_name: str = ""
    device_path: str = ""
    vendor_id: int | None = None
    product_id: int | None = None
    axis_profile: str = ""
    duration_seconds: float = 0.0
    overall: TestStatus = TestStatus.NOT_TESTED
    tests: dict[str, TestStatus] = field(default_factory=dict)
    left_stick: StickTestResult = field(default_factory=StickTestResult)
    right_stick: StickTestResult = field(default_factory=StickTestResult)
    lt: TriggerTestResult = field(default_factory=TriggerTestResult)
    rt: TriggerTestResult = field(default_factory=TriggerTestResult)
    buttons: ButtonsTestResult = field(default_factory=ButtonsTestResult)
    event_rate: EventRateResult = field(default_factory=EventRateResult)
    issues: list[str] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)
    score: int = 0
    disclaimer: str = ""
    thresholds: dict[str, float] = field(default_factory=dict)

    def finalize(self) -> None:
        self.overall = overall_from_tests(self.tests)
        self.issues = []
        self.notes = []
        for res in (self.left_stick, self.right_stick):
            self.issues.extend(res.issues)
            if res.status not in (TestStatus.NOT_TESTED, TestStatus.NOT_SUPPORTED) and looks_like_square_gate(
                res.circularity_pct
            ):
                self.notes.append("circularity.square_ok")
        self.issues.extend(self.lt.issues)
        self.issues.extend(self.rt.issues)
        self.issues.extend(self.buttons.issues)
        seen: set[str] = set()
        unique_notes: list[str] = []
        for note in self.notes:
            if note not in seen:
                seen.add(note)
                unique_notes.append(note)
        self.notes = unique_notes
        tested = any(
            status not in (TestStatus.NOT_TESTED, TestStatus.NOT_SUPPORTED)
            for status in self.tests.values()
        ) or self.overall != TestStatus.NOT_TESTED
        self.score = compute_score(self) if tested else 0

    def status_label_key(self) -> str:
        mapping = {
            TestStatus.PASS: "status.pass",
            TestStatus.WARN: "status.warn",
            TestStatus.FAIL: "status.fail",
            TestStatus.NOT_TESTED: "status.not_tested",
        }
        return mapping.get(self.overall, "status.not_tested")

    def score_css_class(self) -> str:
        return {
            "good": "score-good",
            "ok": "score-ok",
            "bad": "score-bad",
        }.get(score_tone(self.score), "score-ok")


def button_check_status(item: ButtonCheck) -> TestStatus:
    if item.skipped or not item.pressed:
        return TestStatus.FAIL
    if item.sticky or not item.sensitive or not item.held:
        return TestStatus.WARN
    return TestStatus.PASS


def _bare_issue(issue: str, name: str) -> str:
    if issue.startswith(f"{name}: "):
        return issue[len(name) + 2 :]
    if issue.startswith(f"{name} "):
        return issue[len(name) + 1 :]
    return issue


def _control_lines(name: str, status: TestStatus, issues: list[str], extra: str = "") -> list[str]:
    title = f"{name} — {status.value}"
    if extra:
        title = f"{title} ({extra})"
    lines = [title]
    lines.extend(f"• {_bare_issue(issue, name)}" for issue in issues)
    return lines


def _stick_extra(stick: StickTestResult) -> str:
    from core.i18n import t

    parts = [t("report.drift", pct=f"{stick.drift_pct:.1f}")]
    if stick.circularity_pct > 0:
        parts.append(t("report.gate_shape", pct=f"{stick.circularity_pct:.1f}"))
    return " · ".join(parts)


def _extend_section(lines: list[str], heading: str, blocks: list[list[str]]) -> None:
    if not blocks:
        return
    if lines and lines[-1] != "":
        lines.append("")
    lines.append(heading)
    for index, block in enumerate(blocks):
        if index and (len(blocks[index - 1]) > 1 or len(block) > 1):
            lines.append("")
        lines.extend(block)


def result_lines(report: DiagnosticReport) -> list[str]:
    from core.i18n import t

    lines = [
        t("diag.score", score=report.score),
        f"{t('diag.status')}: {t(report.status_label_key())}",
    ]

    stick_blocks: list[list[str]] = []
    for stick, label in (
        (report.left_stick, t("stick.left")),
        (report.right_stick, t("stick.right")),
    ):
        if stick.status == TestStatus.NOT_TESTED:
            continue
        stick_blocks.append(_control_lines(label, stick.status, stick.issues, _stick_extra(stick)))
    _extend_section(lines, t("diag.sticks"), stick_blocks)

    trigger_blocks: list[list[str]] = []
    for trig, label in ((report.lt, "LT"), (report.rt, "RT")):
        if trig.status == TestStatus.NOT_TESTED:
            continue
        extra = f"{trig.max_value * 100:.0f}%"
        trigger_blocks.append(_control_lines(label, trig.status, trig.issues, extra))
    _extend_section(lines, t("diag.triggers"), trigger_blocks)

    button_blocks: list[list[str]] = []
    if report.buttons.status != TestStatus.NOT_TESTED:
        for item in report.buttons.buttons:
            button_blocks.append(
                _control_lines(item.name, button_check_status(item), item.issues)
            )
    _extend_section(lines, t("diag.buttons"), button_blocks)

    for note in report.notes:
        lines.append("")
        lines.append(t(note))
    return lines

File: src-tauri/src/report.rs
use std::collections::{HashMap, HashSet};

use serde::Serialize;

use crate::i18n::{t, t_vars};
use crate::pad::{APP_NAME, REPORT_SCHEMA, VERSION};
use crate::stats::looks_like_square_gate;
use crate::status::{overall_from_tests, TestStatus};

#[derive(Clone, Debug, Serialize)]
pub struct StickTestResult {
    pub status: TestStatus,
    pub drift_pct: f64,
    pub mean_x: f64,
    pub mean_y: f64,
    pub max_radius: f64,
    pub deadzone_pct: f64,
    pub physical_drift_pct: f64,
    pub circularity_pct: f64,
    pub range_ok: bool,
    pub issues: Vec<String>,
    pub rest_points: Vec<(f64, f64)>,
}

impl Default for StickTestResult {
    fn default() -> Self {
        Self {
            status: TestStatus::NotTested,
            drift_pct: 0.0,
            mean_x: 0.0,
            mean_y: 0.0,
            max_radius: 0.0,
            deadzone_pct: 0.0,
            physical_drift_pct: 0.0,
            circularity_pct: 0.0,
            range_ok: true,
            issues: Vec::new(),
            rest_points: Vec::new(),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct TriggerTestResult {
    pub status: TestStatus,
    pub min_value: f64,
    pub max_value: f64,
    pub spike_count: i32,
    pub returns_to_zero: bool,
    pub issues: Vec<String>,
    pub timeline: Vec<(u64, f64)>,
}

impl Default for TriggerTestResult {
    fn default() -> Self {
        Self {
            status: TestStatus::NotTested,
            min_value: 0.0,
            max_value: 0.0,
            spike_count: 0,
            returns_to_zero: true,
            issues: Vec::new(),
            timeline: Vec::new(),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct ButtonCheck {
    pub index: u8,
    pub name: String,
    pub pressed: bool,
    pub held: bool,
    pub skipped: bool,
    pub sticky: bool,
    pub sensitive: bool,
    pub tap_count: u32,
    pub issues: Vec<String>,
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct ButtonsTestResult {
    pub status: TestStatus,
    pub buttons: Vec<ButtonCheck>,
    pub pressed_count: i32,
    pub held_count: i32,
    pub issues: Vec<String>,
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct EventRateResult {
    pub mean_interval_ms: f64,
    pub median_interval_ms: f64,
    pub max_interval_ms: f64,
    pub estimated_hz: f64,
    pub note: String,
}

#[derive(Clone, Debug, Serialize)]
pub struct DiagnosticReport {
    pub schema_version: String,
    pub app_name: String,
    pub app_version: String,
    pub created_at: String,
    pub locale: String,
    pub device_name: String,
    pub device_path: String,
    pub vendor_id: Option<u16>,
    pub product_id: Option<u16>,
    pub axis_profile: String,
    pub duration_seconds: f64,
    pub overall: TestStatus,
    pub tests: HashMap<String, TestStatus>,
    pub left_stick: StickTestResult,
    pub right_stick: StickTestResult,
    pub lt: TriggerTestResult,
    pub rt: TriggerTestResult,
    pub buttons: ButtonsTestResult,
    pub event_rate: EventRateResult,
    pub issues: Vec<String>,
    pub notes: Vec<String>,
    pub score: i32,
    pub disclaimer: String,
    pub thresholds: HashMap<String, f64>,
}

impl DiagnosticReport {
    pub fn new(locale: &str, disclaimer: String, thresholds: HashMap<String, f64>) -> Self {
        Self {
            schema_version: REPORT_SCHEMA.into(),
            app_name: APP_NAME.into(),
            app_version: VERSION.into(),
            created_at: chrono::Utc::now().to_rfc3339(),
            locale: locale.to_string(),
            device_name: String::new(),
            device_path: String::new(),
            vendor_id: None,
            product_id: None,
            axis_profile: String::new(),
            duration_seconds: 0.0,
            overall: TestStatus::NotTested,
            tests: HashMap::new(),
            left_stick: StickTestResult::default(),
            right_stick: StickTestResult::default(),
            lt: TriggerTestResult::default(),
            rt: TriggerTestResult::default(),
            buttons: ButtonsTestResult::default(),
            event_rate: EventRateResult::default(),
            issues: Vec::new(),
            notes: Vec::new(),
            score: 0,
            disclaimer,
            thresholds,
        }
    }

    pub fn finalize(&mut self) {
        self.overall = overall_from_tests(&self.tests);
        self.issues.clear();
        self.notes.clear();
        for res in [&self.left_stick, &self.right_stick] {
            self.issues.extend(res.issues.clone());
            if !matches!(res.status, TestStatus::NotTested | TestStatus::NotSupported)
                && looks_like_square_gate(res.circularity_pct)
            {
                self.notes.push("circularity.square_ok".into());
            }
        }
        self.issues.extend(self.lt.issues.clone());
        self.issues.extend(self.rt.issues.clone());
        self.issues.extend(self.buttons.issues.clone());
        let mut seen = HashSet::new();
        self.notes.retain(|note| seen.insert(note.clone()));
        let tested = self.tests.values().any(|s| {
            !matches!(s, TestStatus::NotTested | TestStatus::NotSupported)
        }) || self.overall != TestStatus::NotTested;
        self.score = if tested { compute_score(self) } else { 0 };
    }

    pub fn status_label_key(&self) -> &'static str {
        match self.overall {
            TestStatus::Pass => "status.pass",
            TestStatus::Warn => "status.warn",
            TestStatus::Fail => "status.fail",
            _ => "status.not_tested",
        }
    }
}

pub fn compute_score(report: &DiagnosticReport) -> i32 {
    let mut score = 10.0;
    for stick in [&report.left_stick, &report.right_stick] {
        match stick.status {
            TestStatus::Fail => score -= 3.0,
            TestStatus::Warn => score -= 1.5,
            _ => {}
        }
    }
    for trig in [&report.lt, &report.rt] {
        match trig.status {
            TestStatus::Fail => score -= 2.0,
            TestStatus::Warn => score -= 0.75,
            _ => {}
        }
    }
    let buttons = &report.buttons;
    if !matches!(buttons.status, TestStatus::NotTested | TestStatus::NotSupported) {
        let missed = buttons
            .buttons
            .iter()
            .filter(|item| !item.pressed || item.skipped)
            .count() as f64;
        let not_held = buttons
            .buttons
            .iter()
            .filter(|item| item.pressed && !item.held && !item.skipped)
            .count() as f64;
        let sticky = buttons.buttons.iter().filter(|item| item.sticky).count() as f64;
        let dull = buttons.buttons.iter().filter(|item| !item.sensitive).count() as f64;
        score -= (missed * 1.5).min(6.0);
        score -= (not_held * 0.5).min(2.0);
        score -= (sticky * 1.0).min(3.0);
        score -= (dull * 1.0).min(3.0);
    }
    score.round().clamp(1.0, 10.0) as i32
}

fn button_check_status(item: &ButtonCheck) -> TestStatus {
    if item.skipped || !item.pressed {
        TestStatus::Fail
    } else if item.sticky || !item.sensitive || !item.held {
        TestStatus::Warn
    } else {
        TestStatus::Pass
    }
}

fn bare_issue(issue: &str, name: &str) -> String {
    let prefix_colon = format!("{name}: ");
    let prefix_space = format!("{name} ");
    if let Some(rest) = issue.strip_prefix(&prefix_colon) {
        rest.to_string()
    } else if let Some(rest) = issue.strip_prefix(&prefix_space) {
        rest.to_string()
    } else {
        issue.to_string()
    }
}

fn control_lines(name: &str, status: TestStatus, issues: &[String], extra: &str) -> Vec<String> {
    let mut title = format!("{} — {}", name, status.as_str());
    if !extra.is_empty() {
        title = format!("{title} ({extra})");
    }
    let mut lines = vec![title];
    lines.extend(issues.iter().map(|issue| format!("• {}", bare_issue(issue, name))));
    lines
}

fn stick_extra(stick: &StickTestResult) -> String {
    let mut parts = vec![t_vars("report.drift", &[("pct", format!("{:.1}", stick.drift_pct))])];
    if stick.circularity_pct > 0.0 {
        parts.push(t_vars(
            "report.gate_shape",
            &[("pct", format!("{:.1}", stick.circularity_pct))],
        ));
    }
    parts.join(" · ")
}

fn extend_section(lines: &mut Vec<String>, heading: String, blocks: &[Vec<String>]) {
    if blocks.is_empty() {
        return;
    }
    if let Some(last) = lines.last() {
        if !last.is_empty() {
            lines.push(String::new());
        }
    }
    lines.push(heading);
    for (index, block) in blocks.iter().enumerate() {
        if index > 0 && (blocks[index - 1].len() > 1 || block.len() > 1) {
            lines.push(String::new());
        }
        lines.extend(block.clone());
    }
}

pub fn result_lines(report: &DiagnosticReport) -> Vec<String> {
    let mut lines = vec![
        t_vars("diag.score", &[("score", report.score.to_string())]),
        format!("{}: {}", t("diag.status"), t(report.status_label_key())),
    ];

    let mut stick_blocks = Vec::new();
    for (stick, label) in [
        (&report.left_stick, t("stick.left")),
        (&report.right_stick, t("stick.right")),
    ] {
        if stick.status == TestStatus::NotTested {
            continue;
        }
        stick_blocks.push(control_lines(
            &label,
            stick.status,
            &stick.issues,
            &stick_extra(stick),
        ));
    }
    extend_section(&mut lines, t("diag.sticks"), &stick_blocks);

    let mut trigger_blocks = Vec::new();
    for (trig, label) in [(&report.lt, "LT"), (&report.rt, "RT")] {
        if trig.status == TestStatus::NotTested {
            continue;
        }
        let extra = format!("{:.0}%", trig.max_value * 100.0);
        trigger_blocks.push(control_lines(label, trig.status, &trig.issues, &extra));
    }
    extend_section(&mut lines, t("diag.triggers"), &trigger_blocks);

    let mut button_blocks = Vec::new();
    if report.buttons.status != TestStatus::NotTested {
        for item in &report.buttons.buttons {
            button_blocks.push(control_lines(
                &item.name,
                button_check_status(item),
                &item.issues,
                "",
            ));
        }
    }
    extend_section(&mut lines, t("diag.buttons"), &button_blocks);

    for note in &report.notes {
        lines.push(String::new());
        lines.push(t(note));
    }
    lines
}
